storage/internal/benchmarks/client_pool.go

```diff
@@ -0,0 +1,196 @@
+			log.Fatalf("storage.NewClient: %v", err)
+		}
+
+		return client
+	},
+}
+
+func initializeClientPools(opts *benchmarkOptions) func() {
+	httpClients = &clientPool{
+		New: func() *storage.Client {
+			client, err := initializeHTTPClient(context.Background(), opts.minWriteSize, opts.maxReadSize, opts.useDefaults)
+			if err != nil {
+				log.Fatalf("initializeHTTPClient: %v", err)
+			}
+
+			return client
+		},
+	}
+
+	gRPCClients = &clientPool{
+		New: func() *storage.Client {
+			client, err := initializeGRPCClient(context.Background(), opts.minWriteSize, opts.maxReadSize, opts.connPoolSize, opts.useDefaults)
+			if err != nil {
+				log.Fatalf("initializeGRPCClient: %v", err)
+			}
+			return client
+		},
+	}
+
+	return func() {
+		for _, c := range httpClients.clients {
+			c.Close()
+		}
+		for _, c := range gRPCClients.clients {
+			c.Close()
+		}
+	}
+}
+
+func canUseClientPool(opts *benchmarkOptions) bool {
+	return opts.useDefaults || (opts.maxReadSize == opts.minReadSize && opts.maxWriteSize == opts.minWriteSize)
+}
+
+func getClient(ctx context.Context, opts *benchmarkOptions, br benchmarkResult) (*storage.Client, func() error, error) {
+	noOp := func() error { return nil }
+	grpc := br.params.api == grpcAPI || br.params.api == directPath
+	if canUseClientPool(opts) {
+		if grpc {
+			c := gRPCClients.Get()
+			return c, func() error { gRPCClients.Put(c); return nil }, nil
+		}
+		c := httpClients.Get()
+		return c, func() error { httpClients.Put(c); return nil }, nil
+	}
+
+	// if necessary, create a client
+	if grpc {
+		c, err := initializeGRPCClient(ctx, br.params.appBufferSize, br.params.appBufferSize, opts.connPoolSize, false)
+		if err != nil {
+			return nil, noOp, fmt.Errorf("initializeGRPCClient: %w", err)
+		}
+		return c, c.Close, nil
+	}
+	c, err := initializeHTTPClient(ctx, br.params.appBufferSize, br.params.appBufferSize, false)
+	if err != nil {
+		return nil, noOp, fmt.Errorf("initializeHTTPClient: %w", err)
+	}
+	return c, c.Close, nil
+}
+
+// mutex on starting a client so that we can set an env variable for GRPC clients
+var clientMu sync.Mutex
+
+func initializeHTTPClient(ctx context.Context, writeBufferSize, readBufferSize int, useDefaults bool) (*storage.Client, error) {
+	if useDefaults {
+		clientMu.Lock()
+		c, err := storage.NewClient(ctx, option.WithCredentialsFile(credentialsFile))
+		clientMu.Unlock()
+		return c, err
+	}
+
+	dialer := &net.Dialer{
+		Timeout:   30 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}
+
+	// These are the default parameters with write and read buffer sizes modified
+	base := &http.Transport{
+		Proxy:                 http.ProxyFromEnvironment,
+		DialContext:           dialer.DialContext,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		WriteBufferSize:       writeBufferSize,
+		ReadBufferSize:        readBufferSize,
+	}
```

Comment:
The HTTP/2 ReadIdleTimeout should also be set at this stage; see https://github.com/googleapis/google-api-go-client/blob/main/transport/http/configure_http2_go116.go